src/compiler/parser.ts

```diff
@@ -1811,7 +1811,7 @@ namespace ts {
         function parseEntityName(allowReservedWords: boolean, diagnosticMessage?: DiagnosticMessage): EntityName {
             let entity: EntityName = parseIdentifier(diagnosticMessage);
             while (parseOptional(SyntaxKind.DotToken)) {
-                const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
+                const node: QualifiedName = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);  // !!!

```

Comment:
@ahejlsberg - this doesn't seem circular to me, it seems sequential. I'm imagining if the first iteration of the loop were unrolled - don't we just need to take the partially resolved type for entity from the first iteration and use it as input for the type of entity for the second iteration? Once you've reached a stable signature (or cycle of signatures) by this method of loop-unrolling, you'd just need the union of all types made in each iteration for the overall type at that location, no?